win32/tests/file_win32_ut/file_win32_ut.c

```diff
@@ -0,0 +1,952 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#ifdef __cplusplus
+#include <cstdlib>
+#include <cstddef>
+#else
+#include <stdlib.h>
+#include <stddef.h>
+#include <stdbool.h>
+#endif
+#include "windows.h"
+#include "azure_macro_utils/macro_utils.h"
+
+void* real_malloc(size_t size)
+{
+    return malloc(size);
+}
+
+void real_free(void* ptr)
+{
+    free(ptr);
+}
+
+#include "testrunnerswitcher.h"
+#include "umock_c/umock_c.h"
+#include "umock_c/umocktypes_windows.h"
+#include "umock_c/umocktypes_bool.h"
+#include "umock_c/umock_c_negative_tests.h"
+
+
+#define ENABLE_MOCKS
+
+#include "azure_c_pal/execution_engine.h"
+#include "azure_c_pal/gballoc.h"
+#include "execution_engine_win32.h"
+#include "mock_file.h"
+
+MOCKABLE_FUNCTION(, void, mock_user_callback, void*, user_context, bool, is_successful);
+
+#undef ENABLE_MOCKS
+
+#include "azure_c_pal/file.h"
+
+static TEST_MUTEX_HANDLE g_testByTest;
+
+MU_DEFINE_ENUM_STRINGS(UMOCK_C_ERROR_CODE, UMOCK_C_ERROR_CODE_VALUES)
+
+TEST_DEFINE_ENUM_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_RESULT)
+IMPLEMENT_UMOCK_C_ENUM_TYPE(FILE_WRITE_ASYNC_RESULT, FILE_WRITE_ASYNC_VALUES)
+
+TEST_DEFINE_ENUM_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_RESULT)
+IMPLEMENT_UMOCK_C_ENUM_TYPE(FILE_READ_ASYNC_RESULT, FILE_READ_ASYNC_VALUES)
+
+static void on_umock_c_error(UMOCK_C_ERROR_CODE error_code)
+{
+    ASSERT_FAIL("umock_c reported error :%" PRI_MU_ENUM "", MU_ENUM_VALUE(UMOCK_C_ERROR_CODE, error_code));
+}
+
+static FILE_HANDLE get_file_handle(const char* filename, PTP_WIN32_IO_CALLBACK* captured_callback)
+{
+    ///arrange
+    EXECUTION_ENGINE_HANDLE execution_engine = (EXECUTION_ENGINE_HANDLE)41;
+    HANDLE handle = (HANDLE)42;
+    PTP_POOL ptp_pool = (PTP_POOL)43;
+    PTP_CALLBACK_ENVIRON captured_pcbe;
+    PTP_CLEANUP_GROUP ptp_cleanup_group = (PTP_CLEANUP_GROUP)44;
+    PTP_IO ptp_io = (PTP_IO)45;
+
+
+    STRICT_EXPECTED_CALL(malloc(IGNORED_ARG));
+    STRICT_EXPECTED_CALL(mock_CreateFileA(filename, GENERIC_READ | GENERIC_WRITE, FILE_SHARE_READ, NULL, OPEN_ALWAYS, FILE_FLAG_OVERLAPPED | FILE_FLAG_WRITE_THROUGH, NULL))
+        .SetReturn(handle);

```

Comment:
>.SetReturn(handle); [](start = 8, length = 19)

When you return a fixed fake for calls like this, the preferred style is to define the variable as a global (`static HANDLE fake_create_file_handle = (HANDLE)42;`). And then in the TEST_SUITE_INIT put `REGISTER_GLOBAL_MOCK_RETURNS(mock_CreateFileA, fake_create_file_handle, NULL);`.

It simplifies this setup code here. Also means you can reference `fake_create_file_handle` elsewhere (like in the destroy tests).

Not strictly required, but I think it's a good cleanup and I'd do it for all these functions that are returning fake fixed values.

Reply:
done

---
In reply to: [456717480](https://github.com/Azure/azure-c-pal/pull/10#discussion_r456717480) [](ancestors = 456717480)